Change the ApiToken rate limit to 23:55 and make it configurable

roesslerj opened the pull request:
As discussed. Title says it all :-)

roesslerj committed:
Change the ApiToken rate limit to 23:55 and make it configurable

ioangut commented on services/ppac/src/main/java/app/coronawarn/datadonation/services/ppac/config/PpacConfiguration.java:
```diff
@@ -169,6 +169,8 @@ public void setOtp(Otp otp) {
     private String ppacIosJwtSigningKey;
     private Integer minDeviceTokenLength;
     private Integer maxDeviceTokenLength;
+    private Integer apiTokenRateLimitMinutes;
```
The validity of the ApiToken it should be in seconds as per new specs.

mlenkeit replied:
Yes, let's go with seconds please. I'll update the "official" spec.

FelixRottler commented on services/ppac/src/main/java/app/coronawarn/datadonation/services/ppac/ios/verification/scenario/ratelimit/ProdPpacIosRateLimitStrategy.java:
```diff
@@ -46,10 +55,10 @@ public void validateForPpa(ApiToken apiToken) {
-      if (hours < VALIDITY_IN_HOURS) {
-        logger.info("Api Token was updated {} hours ago. Api Token can only be used once every {} hours.",
-            hours, VALIDITY_IN_HOURS);
+      long hours = MINUTES.between(lastUsedForPpaUtc, currentTimeUtc);
```
```suggestion
      long diffInMinutes = MINUTES.between(lastUsedForPpaUtc, currentTimeUtc);
```

FelixRottler commented on services/ppac/src/main/java/app/coronawarn/datadonation/services/ppac/ios/verification/scenario/ratelimit/ProdPpacIosRateLimitStrategy.java:
```diff
@@ -46,10 +55,10 @@ public void validateForPpa(ApiToken apiToken) {
-        logger.info("Api Token was updated {} hours ago. Api Token can only be used once every {} hours.",
-            hours, VALIDITY_IN_HOURS);
+      long hours = MINUTES.between(lastUsedForPpaUtc, currentTimeUtc);
+      if (hours < validityInMinutes) {
```
```suggestion
      if (diffInMinutes < validityInMinutes) {
```

FelixRottler commented on services/ppac/src/main/java/app/coronawarn/datadonation/services/ppac/ios/verification/scenario/ratelimit/ProdPpacIosRateLimitStrategy.java:
```diff
@@ -46,10 +55,10 @@ public void validateForPpa(ApiToken apiToken) {
+      long hours = MINUTES.between(lastUsedForPpaUtc, currentTimeUtc);
+      if (hours < validityInMinutes) {
+        logger.info("Api Token was updated {} minutes ago. Api Token can only be used once every {} minutes.",
+            hours, validityInMinutes);
```
```suggestion
            diffInMinutes, validityInMinutes);
```

roesslerj committed:
Use seconds instead of minutes

roesslerj committed:
Add missing setter for apiTokenRateLimitSeconds

roesslerj committed:
Make the failing test pass

roesslerj committed:
Fix flaky TimeUtilsTest.testConversionMethods

roesslerj committed:
Merge branch 'main' into feature/api-rate-limit-configurable

roesslerj committed:
Make TimeUtils same as in CWA-Server

See https://github.com/corona-warn-app/cwa-server/pull/1551/files

roesslerj commented:
Since the TimeUtilsTest#testConversionMethods method kept on failing (being flaky) I adapted the test. Had to add the TimeUtils#setNow method as seen here: https://github.com/corona-warn-app/cwa-server/pull/1551/files

roesslerj committed:
Fixed Checkstyle violations: CustomImportOrder

Import statement for 'java.time.ZoneOffset.UTC' is in the wrong order. Should be in the 'STATIC' group, expecting not assigned imports on this line.
[...]

roesslerj committed:
Also adapt the ProdPpacIosRateLimitStrategy to use the same strategy for time

roesslerj committed:
Fix artefacts from automatic merge...

FelixRottler commented on services/ppac/src/main/java/app/coronawarn/datadonation/services/ppac/ios/verification/scenario/ratelimit/ProdPpacIosRateLimitStrategy.java:
```diff
@@ -58,12 +55,12 @@ public void validateForEdus(ApiToken apiToken) {
-            hours, VALIDITY_IN_HOURS);
+      long seconds = SECONDS.between(lastUsedForPpaUtc, currentTimeUtc);
+      if (seconds < validityInSeconds) {
+        logger.info("Api Token was updated {} seconds ago. Api Token can only be used once every {} seconds.",
```
Even if the check itself is in seconds. For better readability we maybe should still stick to a more concrete timestamp instead of seconds?

roesslerj replied:
Hmm. I know it's less readable, but I would keep the logging the same logic and values as the actual check, to not hide any lurking errors... or how would you actually do it?

FelixRottler replied:
Ok - understandable. I mean we can just extract hours:minutes:seconds from just seconds. So the log would be actually "Api Token was updated 23:56:10 hours ago ..." or "Api Token was updated 23h 56min 10secs ago...."

roesslerj replied:
Good idea. Will do.

FelixRottler commented on services/ppac/src/test/java/app/coronawarn/datadonation/services/ppac/ios/verification/scenario/ratelimit/ProdPpacIosRateLimitStrategyTest.java:
```diff
@@ -6,30 +6,27 @@
+import org.springframework.boot.test.context.SpringBootTest;
 
-@ExtendWith(MockitoExtension.class)
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
```
I would like to keep this as an unit test and not switch to an integration test. We have some other test classes that are explicitly integration tests. Maybe we can add there an additional integration test for this scenario? Would do you think?

roesslerj replied:
Ok. I can play around some more with it.

roesslerj replied:
Done

FelixRottler commented on services/ppac/src/main/java/app/coronawarn/datadonation/services/ppac/ios/verification/scenario/ratelimit/ProdPpacIosRateLimitStrategy.java:
```diff
@@ -21,19 +23,14 @@
+  /**
+   * Constructs a validator instance.
+   */
+  public ProdPpacIosRateLimitStrategy(PpacConfiguration ppacConfiguration) {
```
Maybe we can keep the configuration as a field instead of the integer itself. So we have more flexibility when we need more/different configuration. Also makes testing easier :) What do you think?

roesslerj replied:
We can have it either way. My rationale here was that it makes it easier to understand and immediately see what is used from the configuration. Why would it make testing easier?

FelixRottler replied:
Agree but I had the feeling that when I was testing this locally, either Spring and Mockito had trouble injecting a Mock when the Constructor Parameter is not assigned to a field.

roesslerj replied:
I solved this with a much simplified test, not even using Mockito and Spring...

roesslerj committed:
Keep this as an unit test and not switch to an integration test

roesslerj committed:
Better format the log message to display something like 23:55:00 instead of 86100
